### PR TITLE
Typeahead to select2

### PR DESCRIPTION
**Pull request by orangejenny**
Redoing https://github.com/dimagi/commcare-hq/pull/13731, this time using select2 instead of atwho. Also deprecating the jQuery caret plugin.

Hold off on merge until I can QA a bit more tomorrow with fresh eyes.

@dannyroberts

### CODE REVIEW

File: corehq/apps/style/static/style/ko/knockout_bindings.ko.js [hunk 1]
```diff
@@ -529,7 +502,7 @@ ko.bindingHandlers.select2 = new function(){
 
     this.SOURCE_KEY = "select2-source";
 
-    this.init = function(element, valueAccessor){
+    this.init = function(element, valueAccessor) {
```
Comment (dimagimon) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: 'valueAccessor' is defined but never used (no-unused-vars)

File: corehq/apps/style/static/style/ko/knockout_bindings.ko.js [hunk 2]
```diff
@@ -566,6 +539,73 @@ ko.bindingHandlers.select2 = new function(){
...
+                    return {
+                        id: term,
+                        text: term,
+                    };
+                }
+            },
+        };
+    };
+
+    this.init = function(element, valueAccessor) {
```
Comment (dimagimon) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: 'valueAccessor' is defined but never used (no-unused-vars)

File: corehq/apps/style/static/style/ko/knockout_bindings.ko.js [hunk 3]
```diff
@@ -566,6 +539,73 @@ ko.bindingHandlers.select2 = new function(){
...
+        that._init(element, that.select2Options(element));
+    }
```
Comment (dimagimon) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: Missing semicolon. (semi)

File: corehq/apps/style/static/style/ko/knockout_bindings.ko.js [hunk 4]
```diff
@@ -529,7 +502,23 @@ ko.bindingHandlers.select2 = new function(){
...
+        var newItems = ko.utils.unwrapObservable(valueAccessor()) || [];
+        for (var i = 0; i < newItems.length; i++) {
+            var text = newItems[i].text || newItems[i];
+            var id = newItems[i].id || newItems[i];
+            source.push({id: id, text: text});
+        }
+        return source;
+    };
+
+    this.init = function(element, valueAccessor) {
```
Comment (dimagimon) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: 'valueAccessor' is defined but never used (no-unused-vars)

File: corehq/apps/style/static/style/ko/knockout_bindings.ko.js [hunk 5]
```diff
@@ -546,23 +535,67 @@ ko.bindingHandlers.select2 = new function(){
...
-        // it, because the select2 options are tied to this specific instance.
-        while(source.length > 0) {
-            source.pop();
-        }
-        var newItems = ko.utils.unwrapObservable(valueAccessor()) || [];
-        for (var i = 0; i < newItems.length; i++) {
-            var text = newItems[i].text || newItems[i];
-            var id = newItems[i].id || newItems[i];
-            source.push({id: id, text: text});
+    this.init = function(element, valueAccessor) {
```
Comment (dimagimon) [inline corehq/apps/style/static/style/ko/knockout_bindings.ko.js]: 'valueAccessor' is defined but never used (no-unused-vars)

**Comment by orangejenny** [comment]
Did some more testing on staging, this looks okay except for one jQuery3 problem which I'll fix in another PR.

**Comment by orangejenny** [comment]
@dannyroberts could you review? jQuery 3 is ready to go, but this PR needs to go out first.

File: corehq/apps/reminders/forms.py [hunk 6]
```diff
@@ -677,7 +677,7 @@ def section_start_fields(self):
                     'start_date',
                     placeholder=_("Enter Case Property"),
                     css_class="input-xlarge",
-                    data_bind="typeahead: getAvailableCaseProperties",
+                    data_bind="autocompleteSelect2: getAvailableCaseProperties,"
```
Comment (dannyroberts) [inline corehq/apps/reminders/forms.py]: Did you mean to transpose the `,` and `"`?

File: corehq/apps/reminders/forms.py [hunk 7]
```diff
@@ -723,7 +723,7 @@ def section_recipient(self):
                     InlineField(
                         'recipient_case_match_property',
                         css_class="input-xlarge",
-                        data_bind="typeahead: getAvailableSubcaseProperties",
+                        data_bind="autocompleteSelect2: getAvailableSubcaseProperties,",
```
Comment (dannyroberts) [inline corehq/apps/reminders/forms.py]: here and one below as well. I'm not sure if it breaks anything or not, just flagging.

**Review by dannyroberts — APPROVED** [review]
Looks good to me!

File: corehq/apps/reminders/forms.py (hunk 6 above)
Comment (orangejenny) [inline corehq/apps/reminders/forms.py]: Shouldn't break anything, but wasn't intentional. Reverted.

**Pull request merged**